common/darwin/Classes/AudioUtils.m

```diff
@@ -8,73 +8,111 @@ @implementation AudioUtils
+    for (AVAudioSessionPortDescription *port in rtcSession.session.availableInputs) {
+        if ([port.portType isEqualToString:type]) {
+            inputPort = port;
+            break;
+        }
     }
-  }
-  if (inputPort != nil) {
-    NSError *errOut = nil;
-    [rtcSession lockForConfiguration];
-    [rtcSession setPreferredInput:inputPort error:&errOut];
-    [rtcSession unlockForConfiguration];
-    if(errOut != nil) {
-      return NO;
+    if (inputPort != nil) {
+        NSError *errOut = nil;
+        [rtcSession lockForConfiguration];
+        [rtcSession setPreferredInput:inputPort error:&errOut];
+        [rtcSession unlockForConfiguration];
+        if(errOut != nil) {
+            return NO;
+        }
+        return YES;
     }
-    return YES;
-  }
 #endif
-  return NO;
+    return NO;
 }
 
 + (void)setSpeakerphoneOn:(BOOL)enable {
 #if TARGET_OS_IPHONE
     RTCAudioSession *session = [RTCAudioSession sharedInstance];
+    RTCAudioSessionConfiguration *config = [RTCAudioSessionConfiguration webRTCConfiguration];
     [session lockForConfiguration];
-    [session setCategory:AVAudioSessionCategoryPlayAndRecord
-                  withOptions:enable ? AVAudioSessionCategoryOptionDefaultToSpeaker
-                  :
-                  AVAudioSessionCategoryOptionAllowBluetooth | AVAudioSessionCategoryOptionAllowBluetoothA2DP
-                    error:nil];
+    NSError *error = nil;
+    if(!enable) {
+        [session setCategory:config.category
+                 withOptions:config.categoryOptions
+                       error:&error];
+        [session setPreferredOutputNumberOfChannels:0 error:nil];
+        BOOL success = [session setActive:YES error:&error];
+        if (!success) NSLog(@"Audio session override failed: %@", error);
+        else NSLog(@"AudioSession override via Earpiece/Headset is successful ");
+    } else {
+        BOOL success = [session setCategory:config.category
+                                withOptions:AVAudioSessionCategoryOptionDefaultToSpeaker
+                                      error:&error];
+        [session setPreferredOutputNumberOfChannels:0 error:nil];
+        if (!success)  NSLog(@"Port override failed due to: %@", error);
+        success = [session setActive:YES error:&error];
+        if (!success) NSLog(@"Audio session override failed: %@", error);
+        else NSLog(@"AudioSession override via Loudspeaker is successful ");
+    }
     [session unlockForConfiguration];
 #endif
 }
+
++ (void)updateAudioRoute:(BOOL)speakerOn {
+    RTCAudioSession *session = [RTCAudioSession sharedInstance];
+    NSError *error = nil;
+    [session lockForConfiguration];
+    if(speakerOn) {
+        [session overrideOutputAudioPort:[AVAudioSessionPortBuiltInSpeaker intValue] error: &error];
+        [session.session overrideOutputAudioPort:AVAudioSessionPortOverrideSpeaker error:&error];
+        
+        [session setMode:AVAudioSessionModeVoiceChat error: &error];
```

Comment:
What is the reason for setting session mode here each time?

Actually AVAudioSession provide two modes: [AVAudioSessionModeVoiceChat](https://developer.apple.com/documentation/avfaudio/avaudiosessionmodevoicechat) and [AVAudioSessionModeVideoChat](https://developer.apple.com/documentation/avfaudio/avaudiosessionmodevideochat). And it will be more reasonable to set the according to actually used stream, and not according to speaker on/off.

Reply:
I am on iOS 16, if I don't set mode, it will cause the echo cancellation to be disabled, and the loopback test will whistle